api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractModel.getId(), contractModel.getProposer().getId());
+        return proposerVote != null && proposerVote.isViewerFlaggedComplete();
+    }
+
+    private boolean streamerMarkedContractFailed(@NonNull final ContractModel contractModel) {
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractModel.getId(), contractModel.getStreamer().getId());
+        return streamerVote != null && !streamerVote.isViewerFlaggedComplete();
+    }
+
+    private boolean userCanVoteOnContract(@NonNull final UserModel voter, @NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+
+        final UUID proposerId = contractModel.getProposer().getId();
+        final boolean isVoterContractProposer = voter.getId().equals(proposerId);
+        final boolean proposerHasVoted = (voteModelRepository.findByContractIdAndVoterId(contractId, proposerId) != null);
+
+        final UUID streamerId = contractModel.getStreamer().getId();
+        final boolean isVoterContractStreamer = voter.getId().equals(streamerId);
+        final boolean streamerHasVoted = (voteModelRepository.findByContractIdAndVoterId(contractId, streamerId) == null);
+
+        final boolean isUserAllowedToVote = isVoterContractProposer || isVoterContractStreamer;
+        if (!isUserAllowedToVote) {
+            log.warn("Unauthorized vote detected. User: {} attempted to vote on contractId: {} without having permissions to do so",
+                    voter.getId(),
+                    contractModel.getId());
+        }
+
+        final boolean isProposerAndHasVoted = isVoterContractProposer && proposerHasVoted;
+        final boolean isStreamerAndHasVoted = isVoterContractStreamer && streamerHasVoted;
+        final boolean userHasVoted = isProposerAndHasVoted || isStreamerAndHasVoted;
+        if(userHasVoted) {
+            log.warn("Multiple votes detected. User: {} attempted to vote on contractId: {} more than once",
+                    voter.getId(),
+                    contractModel.getId());
+        }
+        return isUserAllowedToVote && !userHasVoted;
+    }
+
+    private void validateContract(@NonNull final ContractModel contractModel) {
+        if (contractModel.isCommunityContract()) {
+            throw new IllegalArgumentException("Voting logic not yet implemented for community contracts. ContractId " + contractModel.getId());
+        }
+
+        if (!contractModel.isAccepted()) {
+            throw new IllegalArgumentException("Cannot vote on an unaccepted contract. ContractId " + contractModel.getId());
+        }
+
+        if (contractModel.getIsCompleted() != null) {
+            throw new IllegalArgumentException("Cannot vote on a completed contract. ContractId " + contractModel.getId());
+        }
```

Comment:
Add a TODO here for checking isDeclined and isExpired

Reply:
Good callout. done